url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/WebSocketImpl.java

```diff
@@ -0,0 +1,737 @@
+	}
+
+	public void close( InvalidDataException e ) {
+		close( e.getCloseCode(), e.getMessage(), false );
+	}
+
+	/**
+	 * Send Text data to the other end.
+	 * 
+	 * @throws IllegalArgumentException
+	 * @throws NotYetConnectedException
+	 */
+	@Override
+	public void send( String text ) throws WebsocketNotConnectedException {
+		if( text == null )
+			throw new IllegalArgumentException( "Cannot send 'null' data to a WebSocketImpl." );
+		send( draft.createFrames( text, role == Role.CLIENT ) );
+	}
+
+	/**
+	 * Send Binary data (plain bytes) to the other end.
+	 * 
+	 * @throws IllegalArgumentException
+	 * @throws NotYetConnectedException
+	 */
+	@Override
+	public void send( ByteBuffer bytes ) throws IllegalArgumentException , WebsocketNotConnectedException {
+		if( bytes == null )
+			throw new IllegalArgumentException( "Cannot send 'null' data to a WebSocketImpl." );
+		send( draft.createFrames( bytes, role == Role.CLIENT ) );
+	}
+
+	@Override
+	public void send( byte[] bytes ) throws IllegalArgumentException , WebsocketNotConnectedException {
+		send( ByteBuffer.wrap( bytes ) );
+	}
+
+	private void send( Collection<Framedata> frames ) {
+		if( !isOpen() )
+			throw new WebsocketNotConnectedException();
+		for( Framedata f : frames ) {
+			sendFrame( f );
+		}
+	}
+
+	@Override
+	public void sendFragmentedFrame( Opcode op, ByteBuffer buffer, boolean fin ) {
+		send( draft.continuousFrame( op, buffer, fin ) );
+	}
+
+	@Override
+	public void sendFrame( Framedata framedata ) {
+		if( DEBUG )
+			System.out.println( "send frame: " + framedata );
+		write( draft.createBinaryFrame( framedata ) );
+	}
+
+	@Override
+	public boolean hasBufferedData() {
+		return !this.outQueue.isEmpty();
+	}
+
+	private HandshakeState isFlashEdgeCase( ByteBuffer request ) throws IncompleteHandshakeException {
+		request.mark();
+		if( request.limit() > Draft.FLASH_POLICY_REQUEST.length ) {
+			return HandshakeState.NOT_MATCHED;
+		} else if( request.limit() < Draft.FLASH_POLICY_REQUEST.length ) {
+			throw new IncompleteHandshakeException( Draft.FLASH_POLICY_REQUEST.length );
+		} else {
+
+			for( int flash_policy_index = 0 ; request.hasRemaining() ; flash_policy_index++ ) {
+				if( Draft.FLASH_POLICY_REQUEST[ flash_policy_index ] != request.get() ) {
+					request.reset();
+					return HandshakeState.NOT_MATCHED;
+				}
+			}
+			return HandshakeState.MATCHED;
+		}
+	}
+
+	public void startHandshake( ClientHandshakeBuilder handshakedata ) throws InvalidHandshakeException {
+		assert ( readystate != READYSTATE.CONNECTING ) : "shall only be called once";
+
+		// Store the Handshake Request we are about to send
+		this.handshakerequest = draft.postProcessHandshakeRequestAsClient( handshakedata );
+
+		resourceDescriptor = handshakedata.getResourceDescriptor();
+		assert( resourceDescriptor != null );
+		
+		// Notify Listener
+		try {
+			wsl.onWebsocketHandshakeSentAsClient( this, this.handshakerequest );
+		} catch ( InvalidDataException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 629.